### PR TITLE
Simplification of Api Gateway Proxy Lambda creation

### PR DESCRIPTION
**Pull request by SwiftEngineer**
### Removal of `api` package from both the `scalambda-core` and `scalambda-testing` libraries.

This package is what allowed people to be able to create serverless apis without knowing what the heck a "Api Gateway Proxy Request" even is, or how Api Gateway and Lambda are able to connect with one another. I mentioned this change briefly in #29 , but here is some extra details. We have a few internal applications that have had a lot of success using the helpers defined in these packages to both test and develop APIs, and this PR removes all of those helpers and replaces them with... wait for it... NOTHING. They're just plain gone.

**Before anyone starts grabbing for their pitch forks:** I'd like to try and just explain why (while some people are gonna hate on this PR) I still believe it is a move in the right direction.

While the presence of the ApiResource abstract class didn't require users to use it, many did. With many users, came many different use-cases, which meant that we frequently had to iterate on the ApiResource class (and supporting classes for it like ScalambdaApi and the response and request objects). In between these iterations, I think a few users were left feeling a bit helpless, because even though the solution was quite simple, it was buried under a lot of convoluted code inside these helper classes. Take issue #32 for example. Several users are still suffering from this problem, but [look how easy it is to set headers in the Api Gateway Proxy Lambda example function](https://carpe.github.io/scalambda/docs/api/create-api/#defining-each-endpoint).

The fact is, there are many other things that users just don't have easy ways of doing without either jumping deep into scalambda's source code, or reading up on how Api Gateway Proxy Lambdas work in the AWS documentation. Scalambda would need to become as comprehensive as the Play framework in order for us to ever be able to keep up with the demands of users for all these more advanced features.

**TL;DR:** It is my belief that, while `ApiResource` provided pre-baked solutions for a lot of problems, it created a barrier that kept users from getting more familiar with how Api Gateway Proxy Lambdas work. This in turn ended up causing more problems than `ApiResource` solved in the first place. 

### Other Updates

- New way of defining Api Gateway instances that allows you to connect the same lambda function to multiple endpoints (reducing cost and cold start frequency)
- support for PATCH, HEAD, and OPTIONS request handlers
- Fix bad link in documentation site
- Update circe

### CODE REVIEW

File: plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala [hunk 1]
```diff
@@ -31,36 +30,40 @@ object ScalambdaPlugin extends AutoPlugin {
 
   object autoImport extends ScalambaKeys {
 
-    def functionNaming: FunctionNaming.type = FunctionNaming
-    def iamRoleSource: FunctionRoleSource.type = FunctionRoleSource
-    def functionSource: FunctionSource.type = FunctionSource
```
Comment (MrMizz) [inline plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala:36]: i see this red diff, but it still remains as an argument to `scalamda()` ?

File: plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala [hunk 2]
```diff
@@ -103,40 +105,56 @@ object ScalambdaPlugin extends AutoPlugin {
...
+      val apiEndpointSettings = Seq(
+        apiName := apiGatewayInstanceName,
+
+        scalambdaFunctions ++= routes.map(_._2),
+
+        scalambdaApiEndpoints ++= Chain.fromSeq(routes),
+      )
+
+      apiEndpointSettings ++ scalambdaLibs
+    }
```
Comment (MrMizz) [inline plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala:154]: I think we should update the docs to reflect the deprecation of `scalambda` in favor of `apiGatewayDefinition`.
Possibly in `configuration.md`. I can take a swing at it, once I figure out how to resolve in my build.

File: plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala (hunk 1 above)
Comment (MrMizz) [reply plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala:36]: Okay, this is deprecated.

Comment (SwiftEngineer) [reply plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala:36]: Yup yup. I haven't seen any projects that used those three functions (which would make sense as they were undocumented). I did move some of them into the `keys` package though.

File: plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala (hunk 2 above)
Comment (SwiftEngineer) [reply plugin/src/main/scala/io/carpe/scalambda/ScalambdaPlugin.scala:154]: Do you mean `scalambdaEndpoint` ? `scalambda` is still a valid non-deprecated function for defining functions that aren't connected to Api Gateway. Although, I do confess that you have a good point that it feels a bit odd to move the Api Gateway lambdas to this new configuration method, while leaving normal lambdas in the dust.

Maybe we create a `lambdas` setting key that you add lambdas to instead of running `scalambda`? Aside from the benefits of keeping things consistent, maybe it also helps people to write their own helpers for DRYing up their `build.sbt` files? :shrug:
